### PR TITLE
Add overfit sanity check script

### PR DESCRIPTION
**Pull request by ImYangYun**
This PR adds an overfitting test for the SignWritingToPoseDiffusion model. The script verifies that the model can overfit on two extreme synthetic samples: one with all-zero pose targets and another with all-one pose targets. This is used to confirm that the end-to-end inference pipeline can perfectly reconstruct these trivial cases.

- Builds the model with deterministic settings (dropout=0, cond_mask_prob=0, fixed random seeds)
- Prepares two extreme SignWriting samples with targets of all zeros and all ones
- Trains the model on these samples for sufficient epochs until loss approaches zero
- Runs inference on both samples after training, rounding outputs and asserting perfect reconstruction
- Saves and plots the training loss curve as overfit_loss_curve.png
overfit_loss_curve

### CODE REVIEW

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    Create a dummy sample: all-zero or all-one pose, with SignWriting image embedding.
+    """
+    x = torch.full((batch_size, num_keypoints, num_dims, num_past_frames), value, device=device)
+    timesteps = torch.zeros(batch_size, dtype=torch.long, device=device)
+    past_motion = torch.full((batch_size, num_keypoints, num_dims, num_past_frames), value, device=device)
+    pil_img = signwriting_to_clip_image(signwriting_str)
+    sw_img = clip_processor(images=pil_img, return_tensors="pt").pixel_values.to(device)
+    return x, timesteps, past_motion, sw_img
+
+def main():
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:26]: this file should be a proper test file - that means, the functions should be `test_model_overfits` and it should run using pytest.

See this overfit test file for example:
https://github.com/sign-language-processing/transcription/blob/main/text_to_pose/diffusion/tests/model_overfit_test.py#L26

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import matplotlib.pyplot as plt
+from transformers import CLIPProcessor
+
+from signwriting_animation.diffusion.core.models import SignWritingToPoseDiffusion
+from signwriting_evaluation.metrics.clip import signwriting_to_clip_image
+
+# Set random seed for reproducibility
+torch.manual_seed(42)
+if torch.cuda.is_available():
+    torch.cuda.manual_seed_all(42)
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:12]: this is going to repeat in many places, and still not be fully correct.
instead, i recommend 
```py
from lightning import seed_everything

seed_everything(42)
```

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+def make_sample(signwriting_str, value, device, clip_processor,
+                batch_size=1, num_past_frames=10, num_keypoints=21, num_dims=3):
+    """
+    Create a dummy sample: all-zero or all-one pose, with SignWriting image embedding.
+    """
+    x = torch.full((batch_size, num_keypoints, num_dims, num_past_frames), value, device=device)
+    timesteps = torch.zeros(batch_size, dtype=torch.long, device=device)
+    past_motion = torch.full((batch_size, num_keypoints, num_dims, num_past_frames), value, device=device)
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:21]: both `x` which i guess is `future_motion` and `past_motion` are of size `num_past_frames`

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    pil_img = signwriting_to_clip_image(signwriting_str)
+    sw_img = clip_processor(images=pil_img, return_tensors="pt").pixel_values.to(device)
+    return x, timesteps, past_motion, sw_img
+
+def main():
+    # Device and processor setup
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:28]: as a test file, i recommend it always runs on `cpu`, that way it will always be the same locally and in github actions

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    # Model hyperparameters
+    num_keypoints = 21
+    num_dims_per_keypoint = 3
+    num_past_frames = 10
+
+    # Build model (ensure dropout=0 and no randomness)
+    model = SignWritingToPoseDiffusion(
+        num_keypoints=num_keypoints,
+        num_dims_per_keypoint=num_dims_per_keypoint,
+        embedding_arch="openai/clip-vit-base-patch32",
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:40]: duplicate string

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        ("AS18701S1870aS2e734S20500M518x533S1870a489x515S18701482x490S20500508x496S2e734500x468", 1.0)
+    ]
+    samples = []
+    for sw, val in sample_configs:
+        x, timesteps, past_motion, sw_img = make_sample(
+            sw, val, device, clip_processor,
+            num_past_frames=num_past_frames,
+            num_keypoints=num_keypoints,
+            num_dims=num_dims_per_keypoint)
+        samples.append((x, timesteps, past_motion, sw_img, val))
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:65]: no need to unpack the tuple then pack it back

```py
samples = [make_sample(sw, val, device, clip_processor,
            num_past_frames=num_past_frames,
            num_keypoints=num_keypoints,
            num_dims=num_dims_per_keypoint) 
        for sw, val in sample_configs]
```

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            loss = loss_fn(output, target)
+            loss.backward()
+            optimizer.step()
+            epoch_loss += loss.item()
+        avg_loss = epoch_loss / len(samples)
+        losses.append(avg_loss)
+        if epoch % 50 == 0:
+            print(f"Epoch {epoch}, Loss: {avg_loss:.6f}")
+        if epoch % 200 == 0:
+            print("Output min/max:", output.min().item(), output.max().item())
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:87]: i would replace the training loop and the setup with some classic lightning code
```py
import torch
import pytorch_lightning as pl
from torch.utils.data import DataLoader

class LightningModel(pl.LightningModule):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.loss_fn = torch.nn.MSELoss()

    def forward(self, x, timesteps, past_motion, sw_img):
        return self.model(x, timesteps, past_motion, sw_img)

    def training_step(self, batch, batch_idx):
        x, timesteps, past_motion, sw_img, val = batch
        output = self(x, timesteps, past_motion, sw_img)
        target = torch.full_like(output, val)
        loss = self.loss_fn(output, target)
        if batch_idx == 0 and self.current_epoch % 200 == 0:
            self.print(f"Output min/max: {output.min().item()}, {output.max().item()}")
        return loss

    def configure_optimizers(self):
        return optim.Adam(model.parameters(), lr=1e-4)

# Assume `samples` is a list of tuples as in your loop; wrap in a DataLoader so we have batches
dataloader = DataLoader(samples, batch_size=4, shuffle=True)

# Instantiate and train
trainer = pl.Trainer(max_epochs=100, log_every_n_steps=1)
model = LightningModel(model)
trainer.fit(model, dataloader)
```

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    plt.title("Overfit Loss Curve")
+    plt.tight_layout()
+    plt.savefig("overfit_loss_curve.png")
+    plt.close()
+    print("Saved loss curve to overfit_loss_curve.png")
+
+    model.eval()
+    with torch.no_grad():
+        for idx, (x, timesteps, past_motion, sw_img, val) in enumerate(samples):
+            output = model(x, timesteps, past_motion, sw_img)
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:103]: so, this is "kinda" correct - indeed, when asking the model for a single step, it can predict the result. 
But, the correct way to sample from a diffusion model is using https://github.com/AIGAnimation/CAMDM/blob/0c744b84133ca453dcf45aafdfbf49c6ff2841f9/PyTorch/diffusion/gaussian_diffusion.py#L626

Like Rong's work for example (her code is still quite messy though):
https://github.com/sign-language-processing/fluent-pose-synthesis/pull/13/files#diff-13f132104203d37aae4eb9e5c7de469b50a07bd55b5be6a30b02ebe6f385724dR319

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
+import torch
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:1]: this test file tests for 2 things at once:
input of (X signwriting and Y history should return Y)
input of (A signwriting and B history should return B)

Now, how can we tell that both X and Y play a role here? as in, what if the model completely ignores the signwriting and just looks at the history? or the other way around?

For this, I recommend now using 4 examples:
```
("M518x529S14c20481x471S27106503x489", 0, -1),        
("M518x529S14c20481x471S27106503x489", 1, 1),        
("M518x533S1870a489x515S18701482x490", 0, 0),
("M518x533S1870a489x515S18701482x490", 1, 2),
 ```

This way, we can tell for sure that both variables play a role in the prediction

File: signwriting_animation/diffusion/tests/test_overfitting.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        x, timesteps, past_motion, sw_img = make_sample(
+            sw, val, device, clip_processor,
+            num_past_frames=num_past_frames,
+            num_keypoints=num_keypoints,
+            num_dims=num_dims_per_keypoint)
+        samples.append((x, timesteps, past_motion, sw_img, val))
+
+
+    # Training loop
+    num_epochs = 2000
```
Comment (AmitMY) [inline signwriting_animation/diffusion/tests/test_overfitting.py:69]: Does it really require 2000 epochs to converge? this test, after you implement the changes, is going to run every single time in CI. if you can run it in 1000 steps and it still works, or even fewer, then that should be the number. Try lowering the epochs until it doesn't work anymore (after you make all other changes)

**Review by AmitMY — CHANGES_REQUESTED** [review]
great start!

**Review by ImYangYun — COMMENTED** [review]
The main features and improvements are as follows:

- Self-contained Overfit Test: The code provides a minimal pipeline that generates synthetic (toy) data, trains the model to overfit, and evaluates whether the model can perfectly fit the small dataset.
- Lightning Module Abstraction: The training and forward logic are wrapped using a custom LightningModule.
- Tested 100 epoch and successfully converged.
